resources/session04/mashup.py

```diff
@@ -0,0 +1,189 @@
+
+
+def get_score_data(elem):
+    inspection_rows = elem.find_all(is_inspection_data_row)
+    samples = len(inspection_rows)
+    total = 0
+    high_score = 0
+    average = 0
+    for row in inspection_rows:
+        strval = clean_data(row.find_all('td')[2])
+        try:
+            intval = int(strval)
+        except (ValueError, TypeError):
+            samples -= 1
+        else:
+            total += intval
+            high_score = intval if intval > high_score else high_score
+    if samples:
+        average = total/float(samples)
+    data = {
+        u'Average Score': average,
+        u'High Score': high_score,
+        u'Total Inspections': samples
+    }
+    return data
+
+
+def result_generator(sorted_list, count):
+    for rest_data in sorted_list[:count]:
+        yield rest_data
+
+
+
+def result_display(*args):
+    sort_dict = {'highscore': 'High Score',
+                 'average': 'Average Score',
+                 'total': 'Total Inspections'}
+    html = load_inspection_page('inspection_page.html')
+    parsed = parse_source(html)
+    content_col = parsed.find("td", id="contentcol")
+    data_list = restaurant_data_generator(content_col)
+    dict_list = []
+    for data_div in data_list:
+        metadata = extract_restaurant_metadata(data_div)
+        inspection_data = get_score_data(data_div)
+        metadata.update(inspection_data)
+        dict_list.append(metadata)
+    for key in sort_dict:
+        if key in args:
+            dict_list = sorted(dict_list, key=lambda k: k[sort_dict[key]], reverse=True)
+        else:
+            dict_list = dict_list
+    if 'reverse' in args:
+        dict_list.reverse()
+    return dict_list
+
+
+
+
+def get_geojson(result):
+    address = " ".join(result.get('Address', ''))
+    if not address:
+        return None
+    geocoded = geocoder.google(address)
+    geojson = geocoded.geojson
+    inspection_data = {}
+    use_keys = (
+        'Business Name', 'Average Score', 'Total Inspections', 'High Score'
+    )
+    for key, val in result.items():
+        if key not in use_keys:
+            continue
+        if isinstance(val, list):
+            val = " ".join(val)
+        inspection_data[key] = val
+    geojson['properties'] = inspection_data
+    return geojson
+
+
+
+if __name__ == '__main__':
+    count = 10
+    for arg in sys.argv:
+        if arg.isdigit():
+            count = int(arg)
+        if arg == 'display':
+            for rest in result_display(*sys.argv)[:count]:
+                print('\n', rest, type(rest))
+        if arg == 'map':
+            total_result = {'type': 'FeatureCollection', 'features': []}
+            for result in result_generator((result_display(*sys.argv)),count):
+                geojson = get_geojson(result)
+                total_result['features'].append(geojson)
+            print('Map created')
+            with open('my_map.json', 'w') as fh:
+                json.dump(total_result, fh)
```

Comment:
The only problem I have with this is that error handling for incorrect or mis-typed arguments is pretty much non-existent.  Remember that users will do all sorts of things with your scripts that you did not intend.  it's worth spending some time to try to make your scripts robust to those sorts of failings.